src/Workspaces/Core/Desktop/Execution/DesktopReferenceSerializationServiceFactory.cs

```diff
@@ -22,10 +29,62 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
 
         private sealed class Service : AbstractReferenceSerializationService
         {
+            // cache for encoding. 
+            // typical low number, high volumn data cache.
+            private static readonly ConcurrentDictionary<Encoding, byte[]> s_encodingCache = new ConcurrentDictionary<Encoding, byte[]>(concurrencyLevel: 2, capacity: 5);
+
             public Service(ITemporaryStorageService service) : base(service)
             {
             }
 
+            public override void WriteTo(Encoding encoding, ObjectWriter writer, CancellationToken cancellationToken)
+            {
+                if (encoding == null)
+                {
+                    base.WriteTo(encoding, writer, cancellationToken);
+                    return;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                writer.WriteByte(EncodingSerialization);
+
+                byte[] value;
+                if (!s_encodingCache.TryGetValue(encoding, out value))
```

Comment:
but why is this bad? already written this way and I dont see reason to change to lambda one?